compiler/src/iree/compiler/ExternalInterfaces/EncodingExternalModels.cpp

```diff
@@ -0,0 +1,124 @@
+// Copyright 2025 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/compiler/ExternalInterfaces/EncodingExternalModels.h"
+#include "iree/compiler/Dialect/Encoding/IR/EncodingDialect.h"
+#include "iree/compiler/Dialect/Encoding/IR/EncodingOps.h"
+#include "iree/compiler/Dialect/Encoding/IR/EncodingTypes.h"
+#include "llvm/ADT/TypeSwitch.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/IR/Builders.h"
+
+namespace mlir::iree_compiler {
+namespace {
+
+struct ContractionAttrPropagationInterface
+    : public IREE::Encoding::EncodingPropagationAttrInterface::ExternalModel<
+          ContractionAttrPropagationInterface, IREE::Encoding::MatmulKAttr> {
+  bool isPropagable(Attribute attr, Value target) const {
+    auto encoding = cast<IREE::Encoding::MatmulKAttr>(attr);
+    Operation *attachedToOperation = target.getDefiningOp();
+    if (!attachedToOperation) {
+      return false;
+    }
+    return TypeSwitch<Operation *, bool>(attachedToOperation)
```

Comment:
I think we don't need the `attachedToOperation` and the check, because they are covered by this TypeSwitch default case. Can you give it a try?

```suggestion
    return TypeSwitch<Operation *, bool>(target.getDefiningOp())
```

Reply:
It can be a nullptr.

Reply:
This just crashes.

Reply:
I see, thanks for giving a try. I just checked the [documentation](https://github.com/llvm/llvm-project/blob/6a16da75d1d52de2211993e25a6485281defe097/llvm/include/llvm/ADT/TypeSwitch.h#L75-L87) and realized that the implementation does not use `dyn_cast_or_null`, and it surprises me. Then please ignore this comment, sorry about that.

```cpp
/// This class implements a switch-like dispatch statement for a value of 'T'
/// using dyn_cast functionality. Each `Case<T>` takes a callable to be invoked
/// if the root value isa<T>, the callable is invoked with the result of
/// dyn_cast<T>() as a parameter.
///
/// Example:
///  Operation *op = ...;
///  LogicalResult result = TypeSwitch<Operation *, LogicalResult>(op)
///    .Case<ConstantOp>([](ConstantOp op) { ... })
///    .Default([](Operation *op) { ... });
///
template <typename T, typename ResultT = void>
class TypeSwitch : public detail::TypeSwitchBase<TypeSwitch<T, ResultT>, T> {
```